src/main/scala/io/iohk/ethereum/merklepatriciatrie/MerklePatriciaTrie.scala

```diff
@@ -0,0 +1,486 @@
+      new MerklePatriciaTrie(Some(newRootHash),
+        updateNodesInStorage(getRootHash, newRootHash, Some(newRoot), Seq(), Seq(newRoot), dataSource, hashFn),
+        hashFn)
+    }
+  }
+
+  /**
+    * This function deletes a (key-value) pair from the trie. If no (key-value) pair exists with the passed trie then there's no effect on it.
+    *
+    * @param key
+    * @return New trie with the (key-value) pair associated with the key passed deleted from the trie.
+    * @throws MPTException if there is any inconsistency in how the trie is build.
+    */
+  def remove(key: K): MerklePatriciaTrie[K, V] = {
+    rootHash map { rootId =>
+      val keyNibbles = HexPrefix.bytesToNibbles(bytes = kSerializer.toBytes(key))
+      val root = getNode(rootId, dataSource)
+      remove(root, keyNibbles) match {
+        case NodeRemoveResult(true, Some(newRoot), nodesToRemoveFromStorage, nodesToUpdateInStorage) =>
+          val newRootHash = newRoot.hash
+          val afterDeleteDataSource = updateNodesInStorage(
+            previousRootHash = getRootHash,
+            newRootHash = newRootHash,
+            newRoot = Some(newRoot),
+            toRemove = nodesToRemoveFromStorage,
+            toUpdate = nodesToUpdateInStorage,
+            dataSource = dataSource,
+            hashFn = hashFn)
+          new MerklePatriciaTrie(Some(newRootHash), afterDeleteDataSource, hashFn)(kSerializer, vSerializer)
+        case NodeRemoveResult(true, None, nodesToRemoveFromStorage, nodesToUpdateInStorage) =>
+          val afterDeleteDataSource = updateNodesInStorage(
+            previousRootHash = getRootHash,
+            newRootHash = EmptyTrieHash,
+            newRoot = None,
+            toRemove = nodesToRemoveFromStorage,
+            toUpdate = nodesToUpdateInStorage,
+            dataSource = dataSource,
+            hashFn = hashFn)
+          new MerklePatriciaTrie(None, afterDeleteDataSource, hashFn)(kSerializer, vSerializer)
+        case NodeRemoveResult(false, _, _, _) => this
+      }
+    } getOrElse {
+      this
+    }
+  }
+
+  @tailrec
+  private def get(node: Node, searchKey: Array[Byte]): Option[Array[Byte]] = node match {
+    case LeafNode(key, value, _) =>
+      if (key sameElements searchKey) Some(value) else None
+    case extNode@ExtensionNode(sharedKey, next, _) =>
+      val (commonKey, remainingKey) = searchKey.splitAt(sharedKey.length)
+      if (searchKey.length >= sharedKey.length && (sharedKey sameElements commonKey)) {
+        val nextNode = getNextNode(extNode, dataSource)
+        get(nextNode, remainingKey)
+      }
+      else None
+    case branch@BranchNode(children, terminator, _) =>
+      if (searchKey.isEmpty) terminator
+      else getChild(branch, searchKey(0), dataSource) match {
+        case Some(child) => get(child, searchKey.slice(1, searchKey.length))
+        case None => None
+      }
+  }
+
+  private def put(node: Node, searchKey: Array[Byte], value: Array[Byte]): NodeInsertResult = {
```

Comment:
Do you think it will be clear enough to have the code split into 3 function, one per node type? Python's code takes advantage of working with lists and not node types to remove some duplicated code lines but that's not possible in our case.